crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -987,191 +987,372 @@ impl<'src> Parser<'src> {
-                    TokenKind::ColonEqual if paren_nesting == 1 => {
-                        treat_it_as_expr = true;
-                        ignore_comma_check = true;
-                        has_seen_colon_equal = true;
-                    }
-                    // Check for starred expressions inside an 1-level nested parens,
-                    // e.g. `with (a, *b): ...`
-                    TokenKind::Star if paren_nesting == 1 && !LITERAL_SET.contains(prev_token) => {
-                        treat_it_as_expr = true;
-                        ignore_comma_check = true;
-                        has_seen_star = true;
-                    }
-                    // Check for `as` keyword outside parens
-                    TokenKind::As => {
-                        treat_it_as_expr = paren_nesting == 0;
-                        ignore_comma_check = true;
-                    }
-                    TokenKind::Comma if !ignore_comma_check => {
-                        // If the comma is outside of parens, treat it as an expression
-                        // if we've seen `(` and `)`.
-                        if paren_nesting == 0 {
-                            treat_it_as_expr = has_seen_lpar && has_seen_rpar;
-                        } else if !has_seen_star && !has_seen_colon_equal {
-                            treat_it_as_expr = false;
-                        }
-                    }
-                    TokenKind::Colon | TokenKind::Newline => break,
-                    _ => {}
-                }
+        let with_item_kind = if self.eat(TokenKind::Lpar) {
+            self.parse_parenthesized_with_items(start, &mut items)
+        } else {
+            WithItemKind::Unparenthesized
+        };
 
-                index += 1;
-                prev_token = self.peek_nth(index);
+        if with_item_kind.is_parenthesized_expression() {
+            // The trailing comma is optional because (1) they aren't allowed in parenthesized
+            // expression context and, (2) We need to raise the correct error if they're present.
+            //
+            // Consider the following three examples:
+            //
+            // ```python
+            // with (item1, item2): ...  # (1)
+            // with (item1, item2),: ...  # (2)
+            // with (item1, item2), item3,: ...  # (3)
+            // ```
+            //
+            // Here, (1) is valid and represents a parenthesized with items while (2) and (3)
+            // are invalid as they are parenthesized expression. Example (3) will raise an error
+            // stating that a trailing comma isn't allowed, while (2) will raise the following
+            // error.
+            //
+            // The reason that (2) expects an expression is because if it raised an error
+            // similar to (3), we would be suggesting to remove the trailing comma, which would
+            // make it a parenthesized with items. This would contradict our original assumption.
```

Comment:
The original assumption being that it's a parenthesized expression but we're suggesting to convert it into a parenthesized with items. We should raise an error based on our assumption because the _actual_ code could go either way.